### PR TITLE
[Merged by Bors] - feat: prove Leibniz π series with Abel's limit theorem

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
This resolves the Zulip comment [here](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Bounds.20on.20alternating.20convergent.20series/near/416938975).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/SpecificLimits/Normed.lean [hunk 1]
```diff
@@ -631,7 +631,15 @@ variable {b : ℝ} {f : ℕ → ℝ} {z : ℕ → E}
 /-- **Dirichlet's Test** for monotone sequences. -/
 theorem Monotone.cauchySeq_series_mul_of_tendsto_zero_of_bounded (hfa : Monotone f)
     (hf0 : Tendsto f atTop (𝓝 0)) (hgb : ∀ n, ‖∑ i in range n, z i‖ ≤ b) :
-    CauchySeq fun n ↦ ∑ i in range (n + 1), f i • z i := by
+    CauchySeq fun n ↦ ∑ i in range n, f i • z i := by
+  suffices CauchySeq fun n ↦ ∑ i in range (n + 1), f i • z i by
```
Comment (MichaelStollBayreuth) [inline Mathlib/Analysis/SpecificLimits/Normed.lean:635]: This should be easier using [CauchySeq.comp_tendsto](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/Cauchy.html#CauchySeq.comp_tendsto).

Anyway, I would assume that it should be possible to modify the previous proof to give the result directly (but I didn't check myself).

Comment (MichaelStollBayreuth) [reply Mathlib/Analysis/SpecificLimits/Normed.lean:635]: Maybe not...

To go via [CauchySeq.comp_tendsto](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/Cauchy.html#CauchySeq.comp_tendsto), one would need a congruence lemma that says that when two sequences agree eventually, one is Cauchy iff the other one is. But I can't find this in Mathlib.

Comment (MichaelStollBayreuth) [reply Mathlib/Analysis/SpecificLimits/Normed.lean:635]: (The problem is that `n - 1 + 1 = n` does not hold for all `n`.)

Comment (MichaelStollBayreuth) [reply Mathlib/Analysis/SpecificLimits/Normed.lean:635]: So I would suggest that you add a general statement
```lean
lemma CauchySeq.shift {α : Type*} [uniformSpace : UniformSpace α]  (u : ℕ → α) (k : ℕ) :+1: 
    CauchySeq (fun n ↦ u (n + k)) ↔ CauchySeq u := ...
```
say in `Mathlib.Topology.UniformSpace.Cauchy` and use that in the proof.

Comment (Parcly-Taxel) [reply Mathlib/Analysis/SpecificLimits/Normed.lean]: Done.

File: Mathlib/Topology/UniformSpace/Cauchy.lean [hunk 2]
```diff
@@ -310,6 +310,17 @@ theorem tendsto_nhds_of_cauchySeq_of_subseq [Preorder β] {u : β → α} (hu :
...
+/-- Any shift of a Cauchy sequence is also a Cauchy sequence. -/
+theorem cauchySeq_shift {u : ℕ → α} (k : ℕ) : CauchySeq u ↔ CauchySeq fun n ↦ u (n + k) := by
+  constructor <;> intro h
+  · exact h.comp_tendsto (tendsto_add_atTop_nat k)
+  · rw [cauchySeq_iff] at h ⊢
+    intro V mV
+    obtain ⟨N, h⟩ := h V mV
+    use N + k
+    intro a _ b _
+    convert h (a - k) (by omega) (b - k) (by omega) <;> omega
```
Comment (MichaelStollBayreuth) [inline Mathlib/Topology/UniformSpace/Cauchy.lean:322]:
```suggestion
    intro a ha b hb
    convert h (a - k) (Nat.le_sub_of_add_le ha) (b - k) (Nat.le_sub_of_add_le hb) <;> omega
```
`omega` is a bit overkill for the first two side goals.

Comment (Parcly-Taxel) [reply Mathlib/Topology/UniformSpace/Cauchy.lean]: Done.

**Review by MichaelStollBayreuth — APPROVED** [review]
This looks good to me now, and I'll exercise my newly obtained powers to maintainer merge :smile:

File: Mathlib/Topology/UniformSpace/Cauchy.lean [hunk 3]
```diff
@@ -310,6 +310,17 @@ theorem tendsto_nhds_of_cauchySeq_of_subseq [Preorder β] {u : β → α} (hu :
   le_nhds_of_cauchy_adhp hu (mapClusterPt_of_comp hf ha)
 #align tendsto_nhds_of_cauchy_seq_of_subseq tendsto_nhds_of_cauchySeq_of_subseq
 
+/-- Any shift of a Cauchy sequence is also a Cauchy sequence. -/
+theorem cauchySeq_shift {u : ℕ → α} (k : ℕ) : CauchySeq u ↔ CauchySeq fun n ↦ u (n + k) := by
```
Comment (sgouezel) [inline Mathlib/Topology/UniformSpace/Cauchy.lean:314]: Could you exchange the two sides of the iff? A general rule of thumb is that the more complicated expression should be on the left of the iff.

Comment (Parcly-Taxel) [reply Mathlib/Topology/UniformSpace/Cauchy.lean]: Done.

**Comment by sgouezel** [comment]
bors r+
Thanks!

**Pull request closed without merging**